### DISCUSSION TITLE
How do I run/debug package tests in VSCode? (And is kitchen sink broken?)

### DISCUSSION THREAD

**Discussion by jtlapp** [vercel/turborepo#3680] [Help] [answered]
I've been unable to figure out how to run and debug tests of monorepo packages from within VSCode, by clicking the "Run" and "Debug" links above a test. My tests work fine when run from the command line.

In the packages directory, the errors I'm getting suggest that VSCode is not pulling in `tsconfig.json`:

```
Support for the experimental syntax 'decorators' isn't currently enabled
```

In the apps directory, it seems that the TypeScript isn't being transpiled:

```
    SyntaxError: /Users/joe/repos/fieldzoo/apps/installer/src/installer.spec.ts: Unexpected token, expected "from" (3:12)

      1 | import { execSync } from "child_process";
      2 | import { join } from "path";
      3 | import type { Kysely } from "kysely";
```

The above scenarios play out regardless of whether the package imports other packages of the monorepo.

I thought I'd see how the [kitchen sink example](https://github.com/vercel/turbo/tree/main/examples/kitchen-sink) does the job, but a degit and then `pnpm install` did not yield buildable code:

```
logger:build: src/__tests__/log.test.ts(3,1): error TS2304: Cannot find name 'jest'.
logger:build: src/__tests__/log.test.ts(3,12): error TS2304: Cannot find name 'global'.
logger:build: src/__tests__/log.test.ts(5,1): error TS2582: Cannot find name 'describe'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha`.
logger:build: src/__tests__/log.test.ts(6,3): error TS2582: Cannot find name 'it'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha`.
logger:build: src/__tests__/log.test.ts(8,5): error TS2304: Cannot find name 'expect'.
logger:build: src/__tests__/log.test.ts(8,12): error TS2584: Cannot find name 'console'. Do you need to change your target library? Try changing the 'lib' compiler option to include 'dom'.
logger:build: src/index.ts(2,3): error TS2584: Cannot find name 'console'. Do you need to change your target library? Try changing the 'lib' compiler option to include 'dom'.
logger:build: tsconfig.json(9,14): error TS6053: File 'tsconfig/base.json' not found.
logger:build:  ELIFECYCLE  Command failed with exit code 2.
logger:build:  WARN   Local package.json exists, but node_modules missing, did you mean to install?
logger:build: ERROR: command finished with error: command (/Users/joe/repos/kitchensink/packages/logger) pnpm r
```

When I open the kitchen sink example in VSCode, I get bazillions of errors, even after reloading the window:

![Screen Shot 2023-02-07 at 15 52 42](https://user-images.githubusercontent.com/138133/217363097-b70c08c5-f4c7-4973-9f7f-5cf662d24da7.png)

The obvious things to look for in the kitchen sink example did not apply: `@types/node` and `@types/jest` are both installed. VSCode is configured to auto-detect the package manager. `node_modules` exists and is populated in the kitchen sink root directory and in each app's directory.

VSCode supports [multi-root workspaces](https://code.visualstudio.com/docs/editor/multi-root-workspaces), which I created for a few packages, but the resulting individual VSCode workspaces aren't recognizing the root `node_modules` for root-installed packages. Maybe the solution is to never install packages at the root level?

**UPDATE**:

The following `.vscode/launch.json` allows me to successfully run my tests via the launcher menu, but boy is this not scalable:

```json
{
  "configurations": [
    {
      "type": "node",
      "name": "test installer",
      "request": "launch",
      "program": "${workspaceFolder}/node_modules/jest/bin/jest",
      "args": ["./src/installer.spec.ts", "--runInBand"],
      "cwd": "${workspaceFolder}/apps/installer",
      "console": "integratedTerminal",
      "internalConsoleOptions": "neverOpen"
    },
    {
      "type": "node",
      "name": "test validators",
      "request": "launch",
      "program": "${workspaceFolder}/node_modules/jest/bin/jest",
      "args": ["./src/lib/InRange.test.ts", "--runInBand"],
      "cwd": "${workspaceFolder}/libs/validators",
      "console": "integratedTerminal",
      "internalConsoleOptions": "neverOpen"
    }
  ]
}
```

**Comment by jtlapp** [comment]
I've found two working solutions so far:

1. Only install packages at the package level, not the root level, and create a [multi-root workspaces file](https://code.visualstudio.com/docs/editor/multi-root-workspaces) listing all packages.
2. Only install packages at the root level, but symbolically link a `node_modules` directory in the package to the root `node_modules`, and create a [multi-root workspaces file](https://code.visualstudio.com/docs/editor/multi-root-workspaces) listing all packages.

Each package can take its own approach. Both approaches require opening a workspace file from VSCode and then switching among packages.

These solutions are for yarn. I'm planning to see how things go with pnpm.

**UPDATE**: A drawback of the symbolic link is that `.gitignore` ignores `node_modules` directories, resulting in these symbolic links not getting checked in. `.gitignore` might have to indicate exactly which `node_modules` directories to ignore.

**Comment by Zamiell** [comment]
pnpm doesn't work great with monorepos: https://github.com/pnpm/pnpm/issues/6030

**Reply by jtlapp** [reply]
Thanks for the heads up.

**Comment by jtlapp** [comment, marked as answer]
I now have testing/debugging working from within VSCode despite putting all dependencies in the root.

I solved the problem by changing the VSCode extension I was using to run Jest tests. The [vscode-jest](https://github.com/jest-community/vscode-jest) extension does not require a local `node_modules`, while the [vscode-jest-runner](https://github.com/firsttris/vscode-jest-runner) extension does. If you're going to install all dependencies at the root level, you'll need to use vscode-jest.

However, both still require that you set up a a VSCode `project.code-workspace` file and open VSCode in workspaces mode.

Gosh, it would have helped for someone to chime in saying, "Hey, I have this working," so we could have compared setups.

**UPDATE**: Except I am unable to individually debug tests using vscode-jest on a Mac. It seems there is no end to the tooling trouble when you attempt to go with root-level dependencies.

**Comment by Suniron** [comment]
In my project, i had to install jest on root `package.json` and put this config in `.vscode/settings.json`: 

```
  "jest.autoRun": {
    "watch": false,
    "onSave": "test-file",
    "onStartup": ["all-tests"]
  },
  "jest.jestCommandLine": "./node_modules/jest/bin/jest.js",
```

I think it is also possible to start jest from my `packages/configs/jest-preset/` folder.

I don't need to use workspace mode.

**Reply by jtlapp** [reply]
I eventually figured out that Turborepo requires that each workspace individually manage its dependencies. Under this structure, there is no issue running Jest.

I also learned that this is the wrong place to be asking questions; the experts just aren't here. But they are quite responsive and helpful on the Discord channel.
